src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -96,43 +96,39 @@ namespace ts.codefix {
         return constIdentifiers;
     }
 
-    function findLastDotThens(func: FunctionLikeDeclaration, checker: TypeChecker): Map<boolean> {
+
+    function getAllCallbacksToReturn(func: FunctionLikeDeclaration, checker: TypeChecker): Map<true> {
         if (!func.body) {
-            return createMap();
+            return createMap<true>();
         }
 
-        // maps nodes to boolean - true indicates that there is another .then() in the callback chain
-        const lastDotThen: Map<boolean> = createMap();
+        const setOfCallbacksToReturn: Map<true> = createMap<true>();
 
         forEachChild(func.body, function visit(node: Node) {
-            const nodeType = checker.getTypeAtLocation(node);
-            if (isCallExpression(node) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(node, "then")) {
-                // false - there is no following .then() in the callback chain
-                lastDotThen.set(getNodeId(node).toString(), false);
-
-                forEachChild(node, function checkChildren(child: Node) {
-                    forEachChild(child, checkChildren);
-
-                    const nodeType = checker.getTypeAtLocation(child);
-                    if (isCallExpression(child) && nodeType && !!checker.getPromisedTypeOfPromise(nodeType) && hasPropertyAccessExpressionWithName(child, "then")
-                        && child.parent && !isPropertyAccessExpression(child.parent) && lastDotThen.get(getNodeId(child.parent).toString()) === false) {
-
-                        // false - there is no following .then() in the callback chain
-                        lastDotThen.set(getNodeId(child).toString(), false);
-                    }
-                    else if (isCallExpression(child) || isIdentifier(child) && nodeType && checker.getPromisedTypeOfPromise(nodeType)) {
-
-                        // true - there is a following .then() in the callback chain
-                        lastDotThen.set(getNodeId(child).toString(), true);
-                    }
-                });
+            if (isLastCallbackOfName(node, checker, "then")) {
+                setOfCallbacksToReturn.set(getNodeId(node).toString(), true);
+                forEach((<CallExpression>node).arguments, visit);
+            }
+            else if (isLastCallbackOfName(node, checker, "catch")) {
+                setOfCallbacksToReturn.set(getNodeId(node).toString(), true);

```

Comment:
Offline, we discussed that all catches not followed by "then" are included in the result set.  Given that there are only two interesting children, I'd actually just visit each one explicitly - that way you can have a separate comment for each.